server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,284 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.health.node;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.coordination.StableMasterHealthIndicatorService;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.health.Diagnosis;
+import org.elasticsearch.health.HealthIndicatorDetails;
+import org.elasticsearch.health.HealthIndicatorImpact;
+import org.elasticsearch.health.HealthIndicatorResult;
+import org.elasticsearch.health.HealthIndicatorService;
+import org.elasticsearch.health.HealthStatus;
+import org.elasticsearch.health.ImpactArea;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class DiskHealthIndicatorService implements HealthIndicatorService {
+    public static final String NAME = "disk";
+
+    private static final Logger logger = LogManager.getLogger(DiskHealthIndicatorService.class);
+
+    private final ClusterService clusterService;
+
+    private static final HealthIndicatorImpact RED_HEALTH_AFFECTED_INDICES_INGEST_IMPACT = new HealthIndicatorImpact(
+        1,
+        "Cannot insert or update documents in the affected indices.",
+        List.of(ImpactArea.INGEST)
+    );
+
+    private static final HealthIndicatorImpact YELLOW_HEALTH_AFFECTED_INDICES_INGEST_IMPACT = new HealthIndicatorImpact(
+        1,
+        "May not be able to insert or update documents in the affected indices.",
+        List.of(ImpactArea.INGEST)
+    );
+
+    private static final HealthIndicatorImpact HEALTH_INGEST_NODES_PIPELINES_IMPACT = new HealthIndicatorImpact(
+        2,
+        "Cannot run ingest pipelines",
+        List.of(ImpactArea.INGEST)
+    );
+
+    private static final HealthIndicatorImpact HEALTH_INGEST_NODES_MONITORING_IMPACT = new HealthIndicatorImpact(
+        2,
+        "Stack monitoring will not run",
+        List.of(ImpactArea.DEPLOYMENT_MANAGEMENT)
+    );
+
+    private static final HealthIndicatorImpact HEALTH_ML_NODES_IMPACT = new HealthIndicatorImpact(
+        2,
+        "Cannot run machine learning jobs",
+        List.of(ImpactArea.ANALYTICS)
+    );
+
+    private static final HealthIndicatorImpact HEALTH_REMOTE_CLUSTER_CLIENT_NODES_IMPACT = new HealthIndicatorImpact(
+        2,
+        "Cannot run cross cluster search",
```

Comment:
I might have missed this... Why is cross cluster search affected by disk?

Reply:
I didn't actually test it, but that was my interpretation from reading https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html#node-roles:
```
Cross-cluster search and cross-cluster replication require the remote_cluster_client role.
```
My assumption was that if cross-cluster search requires a node with the remote_cluster_client role, and if the node with the remote_cluster_client role was out of disk space it would be incapacitated, and so cross-cluster search would not work. Maybe I'm wrong.